### PR TITLE
Added code editor to injection interface

### LINKED ISSUE #4995: Code Injection: Line numbers and syntax highlighting
We currently have the lovely [code injection](#1993) feature, which is accessible via the labs page.

In order to get this ready to be fully shipped and no longer classed as a 'labs' feature, we want to upgrade the textarea to have nice syntax highlighting and line numbers for the HTML/JS/CSS code that will likely be pasted in.

There are three main things to do to complete this:

1. decide on what editor library to use to add the new functionality, this could be CodeMirror or suggestions are welcome (CodeMirror is being dropped from the main Ghost editor, so there's no longer a requirement that it be that) 
2. make sure the syntax highlighting and line numbers are styled to look like they fit in the Ghost admin, this might take a bit of toing and froing with the core team 
3. wrap the fancy textarea functionality up into a little ember component so we can reuse it elsewhere in the admin

This might be a bit of a punt for a `beginner` issue, but I'm confident it suits someone new to the Ghost codebase as it's mostly a wiring job :)

### PR DESCRIPTION
**Pull request by acburdine**
This PR brings syntax highlighting and line numbering to the Code-Injection settings interface

closes #4995

### CODE REVIEW

**Comment by acburdine** [comment]
All my tests passed locally. I think something is still up with the travis setup....

**Comment by ErisDS** [comment]
Awesome, I'm so excited to see a PR for this land! Looks like it's been implemented really, really neatly. The tests were due to GitHub issues I believe, but it's all green now :)

Gonna cc @JohnONolan to take a peek at this :eyes: 

Here's a quick screengrab of how it looks when populated:


Two questions that come to mind for me:

1) Do we need to have the custom vendor theme code? Is there not a way to import a theme with bower and perhaps one that we could use? I only ask because the less code we have in Ghost core the better, always :)
2) Can code folding be disabled? I might be wrong but my gut instinct is that this feature is too advanced and likely to cause confusion for the kinds of users that will use this feature.

**Comment by acburdine** [comment]
@ErisDS I just pushed a change that disables code folding. As to vendor theme code, that file is one I built myself based off of one of the included themes in the ace-builds repo and the color scheme from the current styling. I can pull that file out of the repo and then throw it up in my own personal repository and then import that via bower though.

File: core/client/vendor/theme-ghost.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * An Ace (http://ace.c9.io) Code Editor Theme for Ghost
+ */
+/* global ace */
+
+/* jshint multistr: true */
+ace.define('ace/theme/ghost', ['require', 'exports', 'module', 'ace/lib/dom'], function (require, exports) {
+    
+    exports.isDark = false;
+    exports.cssClass = 'ace-ghost';
+    exports.cssText = '\
```
Comment (novaugust) [inline core/client/vendor/theme-ghost.js]: I believe this'll get es6 transformed if you want to use backticks for literal strings, ie

```js
`
.ace-ghost .ace_gutter {
  background: foo,
  bar: baz
}
`
```

File: core/client/app/components/gh-ace-editor.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    },
+
+    theme: 'ace/theme/ghost',
+
+    update: function () {
+        this.set('value', this.get('editor').getValue());
+    },
+
+    didInsertElement: function () {
+        var editor = window.ace.edit(this.get('id')),
```
Comment (novaugust) [inline core/client/app/components/gh-ace-editor.js]: Would you just use `ace` here, and add a jshint global to the top of the file?

File: core/client/app/components/gh-ace-editor.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        editor.setTheme(this.get('theme'));
+        editor.setOptions(this.get('options'));
+        editor.insert(this.get('value'));
+        editor.on('focus', function () {
+            self.set('isFocused', true);
+        });
+        editor.on('blur', function () {
+            self.set('isFocused', false);
+        });
+        editor.on('change', function () {
```
Comment (novaugust) [inline core/client/app/components/gh-ace-editor.js]: Do we need to `off` these in the `willDestroyElement`?

**Comment by acburdine** [comment]
@novaugust I fixed all the things you suggested. I'm pretty sure I removed all of the editor event listeners

**Comment by novaugust** [comment]
LGTM! Although like Hannah's saying, it'd be nice if we didn't have to add the theme code ourselves... None of the default themes fit well enough? I'd be fine with that file living in its own repo

**Comment by acburdine** [comment]
I looked at a decent amount of theme - although I'm not much of a style person, so I'm not great at determining which colors work. If you want to check out all the ones that come pre-packaged in the ace-builds, go here: http://ajaxorg.github.io/ace-builds/kitchen-sink.html

**Comment by JohnONolan** [comment]
The Github theme should be fine. Also the textareas should be about 680px wide by 300px heigh on desktop, with roughly 24px height for each line of code.

**Comment by acburdine** [comment]
@JohnONolan I fixed the height, but because the form-group max width is 500px, I can't make the editor any larger. Unless I take it out of form-group, which messes with other things.

As to the line-height, ace has a font-size option, but i'm not sure what would be needed for a 24px line-height in terms of font-size

**Comment by acburdine** [comment]
well, since ace is pretty large in terms of file size, closing this for now. Will reopen a new pull request when CodeMirror one is done.

**Pull request closed without merging**